### ISSUE TITLE
Get the parameters from the Dynet ParameterCollection (python)

### ISSUE THREAD

**Issue by alsuhr-c** [clab/dynet#728]
Is there a way to get the set of parameters in my model (ParameterCollection), e.g. as a dictionary that maps the name to the Parameters/LookupParameters object? It doesn't look like there are functions for it in the class I have access to. In _dynet.pyx, in the ParameterCollection class, I can see there's a function ``parameters_list()`` in ``thisptr``, but I can't use ``thisptr``. None of the other ParameterCollection functions seem to do this. 

I can maintain a dictionary myself, but I figured if ParameterCollection probably already keeps track of this, there should be a way to get that from ParameterCollection.

**Comment by neubig** [comment]
I don't believe there's a good way to do this at the moment, but it might be worth adding. Would you mind explaining the use case? Knowing the reason for doing this would help in designing the interface properly.

**Comment by alsuhr-c** [comment]
I would like to investigate the values of the gradients and parameters during training, e.g. show min/max/mean of each parameter's values and gradients after each step or epoch, for debugging purposes. What parameters are in the graph depends on hyperparameters I set (e.g., which kind of attention mechanism I use for a sequence-to-sequence model), so I don't have a few known parameters I can print this information for. Having a name-to-object dictionary of the parameters in the model would make this very easy to implement (again, I can maintain such a dictionary myself; I was just wondering if there was a way to access one that ParameterCollection probably has behind the scenes). 

I know there's a ``check_validity`` flag/function, but I'm debugging more for something like dying ReLUs or other potential issues.

**Comment by hal3** [comment]
I would also like this, unless there's another way to accomplish it. In particular, I'd like to do early stopping. Every time dev error goes down, store away the parameters somewhere, and then load them back in later. I know I can do this by saving to a file and then using .populate(), but it would be really nice if I didn't have to go through a file for that. (Unless there's another way to accomplish this!)

**Comment by blester125** [comment]
I think this would also be useful as a ease of life for getting the size of your model, it would let you do something like this `total_params = sum(np.prod(p.shape()) for p in model.parameters())` you can do this yourself with a list of parameters but then you need to remember to add rnn parameters with something like `parameters.extend(*rnn.get_parameters())`.

**Comment by pmichel31415** [comment]
Fixed by #960
